README.md
# NAME

TestMaker - Create test class classes from your existing lib/

# VERSION

version 1.1

# SYNOPSIS

testMaker

To be run in base dir of project

assuming existing structure:

- lib/My/Class.pm
- lib/My/Class/Stuff.pm

will create

- t/run.t

    Test::Class test runner

- t/tests/My/Test/Class.pm

    Test::Class custom base class

- t/tests/Test/My/Class.pm

    A test class based on lib/My/Class.pm

- t/tests/Test/My/Class/Stuff.pm

    A test class based on lib/My/Class/Stuff.pm

# DESCRIPTION

If you have an existing structure of classes in /lib running this script
will create a skeleton Test::Class test class for each. Within that test
class each of the subs in you lib class will have a test sub created with
a can\_ok test.

The test classes will contain test subs, with a use\_ok for that sub. The
custom base test class, as parent, will make sure each child test class
on startup performs a use\_ok for the lib class you are testing

The script expects to be run in the base of your project (eg the dir
which contains lib/ t/ and so on).

Should a Test::Class test runner not exist one will be created.

A custom base class will also be created if it does not exist, which all
test classes created by this script inherit from.

Once the script has been run, all being well, you will be able to run your
test suite from the base dir of your project using

`prove -lv --merge`

This script performs a check for [Test::Class](http://search.cpan.org/perldoc?Test::Class), [Test::More](http://search.cpan.org/perldoc?Test::More) and
[Class::Data::Inheritable](http://search.cpan.org/perldoc?Class::Data::Inheritable) and will complain at you if they are not
installed. To go ahead and run the script anyway use the -f flag:

`testMaker.pl -f`

# CAVEATS

It currently does not filter out any sub definitions which are contained
within pod sections of your class, and probably never will.

# SEE ALSO

[Test::Class](http://search.cpan.org/perldoc?Test::Class)

[Class::Data::Inheritable](http://search.cpan.org/perldoc?Class::Data::Inheritable)

[Test::More](http://search.cpan.org/perldoc?Test::More)

[Test::Most](http://search.cpan.org/perldoc?Test::Most)

# ACKNOWLEDGEMENTS

Based on the excellent tutorial on Test::Class by Curtis "Ovid" Poe,
<ovid at cpan.org> [found here](http://www.modernperlbooks.com/mt/2009/03/organizing-test-suites-with-testclass.html)

# AUTHOR

Tom Beresford <me at tomberesford dot co dot uk>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Tom Beresford.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for testMaker

1.1    June 12, 2013
    - Finds its own modules on which to act, rather than rather than requiring input data
    - Creates a Test::Class test runner
    - Creates a base test class to inherit startup require che
    - Checks to see that the moduels required to run the tests
    - Added pod

1.0    June 12, 2013
    - Initial version - hacky script


t/tests/My/Test/Class.pm
package My::Test::Class;

use Test::More;
use base qw<Test::Class Class::Data::Inheritable>;

BEGIN {
    __PACKAGE__->mk_classdata('class');
}

sub startup : Tests( startup => 1 ) {
    my $test = shift;

    (my $class = ref $test) =~ s/^Test:://;
    return ok 1, "$class loaded" if $class eq __PACKAGE__;
    use_ok $class or die;
    $test->class($class);
}

1;


t/tests/Test/TestMaker.pm
package Test::TestMaker;

use strict;
use base 'My::Test::Class';
use Test::Most;

sub constructor : Tests() {
    my $test = shift;

    can_ok($test->class(), 'new');
    isa_ok($test->class->new(), 'TestMaker');
}

sub createTests : Tests() {
    my $test = shift;

    can_ok($test->class(), 'createTests');

    TODO: {
        local $TODO = 'Refactor to make this testable';
        ok(0)
    };
}

sub _wanted : Tests() {
    my $test = shift;

    can_ok($test->class(), '_wanted');

    TODO: {
        local $TODO = 'Refactor to make this testable';
        ok(0)
    };
}

sub checkForModules : Tests() {
    my $test = shift;

    can_ok($test->class(), 'checkForModules');

    TODO: {
        local $TODO = 'Refactor to make this testable';
        ok(0)
    };
}

sub createTestBase : Tests() {
    my $test = shift;

    can_ok($test->class(), 'createTestBase');

    TODO: {
        local $TODO = 'Refactor to make this testable';
        ok(0)
    };
}

sub _makeRunFile : Tests() {
    my $test = shift;

    can_ok($test->class(), '_makeRunFile');

    TODO: {
        local $TODO = 'Refactor to make this testable';
        ok(0)
    };
}

sub _makeMyTestClass : Tests() {
    my $test = shift;

    can_ok($test->class(), '_makeMyTestClass');

    TODO: {
        local $TODO = 'Refactor to make this testable';
        ok(0)
    };
}

1;


t/tests/Test/TestMaker/RunFile.pm
package Test::TestMaker::RunFile;

use strict;
use base 'My::Test::Class';
use Test::More;

sub fileData : Tests() {
    my $test = shift;

    can_ok($test->class(), 'fileData');
}

1;


lib/TestMaker.pm
package TestMaker;

# ABSTRACT: Create test class classes from your existing lib/

use strict;
use warnings;
use autodie;
use File::Path qw(make_path);
use File::Find;
use TestMaker::MyTestClass;
use TestMaker::RunFile;
use TestMaker::TestClass;

sub new {
    my $class = shift;
    return bless {}, $class;
}

sub createTests {
    find({wanted => \&_wanted, no_chdir => 1}, 'lib');
}

sub _wanted {
    my $path = $File::Find::name;
    return unless -f $path;

    my $testClass = TestMaker::TestClass->new({libPath => $path});
    $testClass->makeTestFile();
}

sub checkForModules {

    # to require a module from a variable name we need to do some mucking about to
    # translate from :: to /
    for my $module (qw/Test::Class Test::More Class::Data::Inheritable/) {

        # Foo::Bar::Baz => Foo/Bar/Baz.pm
        (my $fn = "$module.pm") =~ s|::|/|g;
        eval {
            require $fn;
            1;
        };
        if ($@) {
            die "$module is required to run the tests created by this script and doesn't seem to be installed
should you wish to run it anyway please use the -f option to override this check
";
        }
    }
}

sub createTestBase {
    make_path('t')               unless -d 't';
    make_path('t/tests/My/Test') unless -d 't/tests/My/Test';

    _makeRunFile();
    _makeMyTestClass();
}

sub _makeRunFile {
    my $runfile = 't/run.t';
    return if -f $runfile;
    open(my $fh, '>', $runfile);
    print $fh TestMaker::RunFile->fileData();
}

sub _makeMyTestClass {
    my $testClass = 't/tests/My/Test/Class.pm';
    return if -f $testClass;
    open(my $fh, '>', $testClass);
    print $fh TestMaker::MyTestClass->fileData();
}

1;

=pod

=head1 SYNOPSIS

testMaker

To be run in base dir of project

assuming existing structure:

=over 4

=item lib/My/Class.pm

=item lib/My/Class/Stuff.pm

=back

will create

=over 4

=item t/run.t

Test::Class test runner

=item t/tests/My/Test/Class.pm

Test::Class custom base class

=item t/tests/Test/My/Class.pm

A test class based on lib/My/Class.pm

=item t/tests/Test/My/Class/Stuff.pm

A test class based on lib/My/Class/Stuff.pm

=back

=head1 DESCRIPTION

If you have an existing structure of classes in /lib running this script
will create a skeleton Test::Class test class for each. Within that test
class each of the subs in you lib class will have a test sub created with
a can_ok test.

The test classes will contain test subs, with a use_ok for that sub. The
custom base test class, as parent, will make sure each child test class
on startup performs a use_ok for the lib class you are testing

The script expects to be run in the base of your project (eg the dir
which contains lib/ t/ and so on).

Should a Test::Class test runner not exist one will be created.

A custom base class will also be created if it does not exist, which all
test classes created by this script inherit from.

Once the script has been run, all being well, you will be able to run your
test suite from the base dir of your project using

C<prove -lv --merge>

This script performs a check for L<Test::Class>, L<Test::More> and
L<Class::Data::Inheritable> and will complain at you if they are not
installed. To go ahead and run the script anyway use the -f flag:

C<testMaker.pl -f>

=head1 CAVEATS

It currently does not filter out any sub definitions which are contained
within pod sections of your class, and probably never will.

=head1 SEE ALSO

L<Test::Class>

L<Class::Data::Inheritable>

L<Test::More>

L<Test::Most>

=head1 ACKNOWLEDGEMENTS

Based on the excellent tutorial on Test::Class by Curtis "Ovid" Poe,
<ovid at cpan.org> L<found here|http://www.modernperlbooks.com/mt/2009/03/organizing-test-suites-with-testclass.html>

=cut


lib/TestMaker/TestSub.pm
package TestMaker::TestSub;

# ABSTRACT: Class to generate test sub definitions

use strict;
use warnings;

sub new {
    my ($class, $args) = @_;

    die "subName not defined" unless $args->{subName};

    my $self = {};
    bless $self, $class;

    $self->subName($args->{subName});
    $self->testSubName($args->{subName});

    return $self;
}

sub subName {
    my ($self, $subName) = @_;

    if (@_ > 1) {
        die "Bad subName" unless $self->_validateSubName($subName);
        chomp($subName);
        $self->{subName} = $subName;
    }
    return $self->{subName};
}

sub testSubName {
    my ($self, $subName) = @_;

    if (@_ > 1) {
        die "Bad subName" unless $self->_validateSubName($subName);
        chomp($subName);
        if ($subName eq 'new') {
            $self->{testSubName} = 'constructor';
        }
        else {
            $self->{testSubName} = $subName;
        }
    }
    return $self->{testSubName};
}

sub _validateSubName {
    my ($self, $subName) = @_;

    unless (defined $subName) {
        warn "Invalid subName: Not defined";
        return 0;
    }

    unless ($subName) {
        warn "Invalid subName: $subName does not evalute to true";
        return 0;
    }

    if ($subName eq 'sub') {
        warn "Invalid subName: cannot use subName sub";
        return 0;
    }

    return 1;
}

sub testSubDefinition {
    my ($self) = @_;

    die "subName not defined" unless $self->subName();

    my $subDef .= q!sub ! . $self->testSubName() . q! : Tests() {
    my $test = shift;

    can_ok($test->class(), '! . $self->subName() . q!');
}

!;
    return $subDef;
}

1;


lib/TestMaker/TestClass.pm
package TestMaker::TestClass;

# ABSTRACT: Class to generate test classes

use strict;
use warnings;
use autodie;
use File::Path qw(make_path);
use TestMaker::TestSub;

sub new {
    my ($class, $args) = @_;

    die "libPath not defined" unless $args->{libPath};

    my $self = {};
    bless $self, $class;

    $self->libPath($args->{libPath});

    return $self;
}

sub libPath {
    my ($self, $libPath) = @_;

    if (@_ > 1) {
        die "Bad libPath" unless $self->_validateLibPath($libPath);
        $self->{libPath} = $libPath;
    }
    return $self->{libPath};
}

sub _validateLibPath {
    my ($self, $libPath) = @_;

    unless (defined $libPath) {
        warn "Invalid libPath: Not defined";
        return 0;
    }

    unless ($libPath) {
        warn "Invalid libPath: $libPath does not evalute to true";
        return 0;
    }

    unless (-f $libPath) {
        warn "Invalid libPath: $libPath not a regular file";
        return 0;
    }

    unless (-r $libPath) {
        warn "Invalid libPath: $libPath cannot be read from";
        return 0;
    }

    return 1;
}

sub makeTestFile {
    my ($self) = @_;

    return if -e $self->_testFilePath();

    $self->_makeTestDir() unless -d $self->_testDir();

    open(my $testFH, '>', $self->_testFilePath());
    print $testFH $self->_makeTestFileContent();
}

sub _testFilePath {
    my ($self) = @_;
    return $self->_testDir() . "/" . $self->_testFile();
}

sub _makeTestDir {
    my ($self) = @_;
    make_path($self->_testDir());
}

sub _testDir {
    my ($self) = @_;
    my $testDir = $self->libPath();
    $testDir =~ s/\.pm//;
    $testDir =~ s!lib/!t/tests/Test/!;
    $testDir =~ s!(.*)/\w+$!$1!;
    return $testDir;
}

sub _testFile {
    my ($self) = @_;
    my $testFile = $self->libPath();
    $testFile =~ s!.*/(\w+\.pm)$!$1!;
    return $testFile;
}

sub _makeTestFileContent {
    my ($self) = @_;

    my $content = $self->_header();

    open(my $libfh, '<', $self->libPath());

    while (<$libfh>) {
        next unless $_ =~ /^\s*sub /;
        my $subName = $self->_tidySubName($_);
        my $sub = TestMaker::TestSub->new({subName => $subName});
        $content .= $sub->testSubDefinition();
    }
    $content .= "1;\n";

    return $content;
}

sub _tidySubName {
    my ($self, $subName) = @_;
    chomp($subName);
    $subName =~ s/^\s*sub\s+([^\(\s\{]+).*/$1/;
    return $subName;
}

sub _testPackageName {
    my ($self) = @_;
    my $testPackageName = $self->libPath();
    $testPackageName =~ s/\.pm//;
    $testPackageName =~ s!lib/(.*)!Test::$1!;
    $testPackageName =~ s!/!::!g;
    return $testPackageName;
}

sub _header {
    my ($self) = @_;

    return "package " . $self->_testPackageName() . ";

use strict;
use base 'My::Test::Class';
use Test::More;

";
}

1;


testMaker.pl
#!/usr/bin/env perl

use strict;
use warnings;
use autodie;
use File::Path qw(make_path);
use File::Find;
use Getopt::Std;

my $opts = {};
getopts('f', $opts);

checkForModules() unless $opts->{f};

createTestBase();

find({wanted => \&wanted, no_chdir => 1}, 'lib');

sub checkForModules {
    # to require a module from a variable name we need to do some mucking about to
    # translate from :: to /
    for my $module (qw/Test::Class Test::More Class::Data::Inheritable/) {
        # Foo::Bar::Baz => Foo/Bar/Baz.pm
        (my $fn = "$module.pm") =~ s|::|/|g;
        eval {
            require $fn;
            1;
        };
        if ($@) {
            die "$module is required to run the tests created by this script and doesn't seem to be installed
should you wish to run it anyway please use the -f option to override this check
";
        }
    }
}

sub createTestBase {
    make_path('t')               unless -d 't';
    make_path('t/tests/My/Test') unless -d 't/tests/My/Test';

    makeRunFile();
    makeMyTestClass();
}

sub makeRunFile {
    my $runfile = 't/run.t';
    return if -f $runfile;

    open(my $fh, '>', $runfile);
    print $fh '#!/usr/bin/env perl

use Test::Class::Load qw<t/tests>;
Test::Class->runtests;';
}

sub makeMyTestClass {
    my $testClass = 't/tests/My/Test/Class.pm';
    return if -f $testClass;

    local $" = '';
    open(my $fh, '>', $testClass);
    my @content = <DATA>;
    print $fh "@content";
}

sub wanted {
    my $path = $File::Find::name;
    return unless -f $path;

    makeTestDir($path);
    makeTestFile($path);
}

sub makeTestDir {
    my ($libFile) = @_;

    my $testDir = testDir($libFile);
    make_path($testDir) unless -d $testDir;
}

sub testDir {
    my ($libFile) = @_;

    $libFile =~ s/\.pm//;
    $libFile =~ s!lib/!t/tests/Test/!;
    $libFile =~ s!(.*)/\w+$!$1!;
    return $libFile;
}

sub makeTestFile {
    my ($libFile) = @_;

    my $testFilePath = testDir($libFile) . "/" . testFile($libFile);
    return if -e $testFilePath;

    my $content = makeTestFileContent($libFile);

    open(my $testFH, '>', $testFilePath);
    print $testFH $content;
}

sub testFile {
    my ($libFile) = @_;

    $libFile =~ s!.*/(\w+\.pm)$!$1!;
    return $libFile;
}

sub makeTestFileContent {
    my ($libFile) = @_;

    my $testPackageName = testPackageName($libFile);
    my $content         = header($testPackageName);

    open(my $libfh, '<', $libFile);

    while (<$libfh>) {
        next unless $_ =~ /^\s*sub /;
        $content .= testSub($_);
    }
    $content .= "1;\n";

    return $content;
}

sub testPackageName {
    my ($libFile) = @_;

    $libFile =~ s/\.pm//;
    $libFile =~ s!lib/(.*)!Test::$1!;
    $libFile =~ s!/!::!g;
    return $libFile;
}

sub header {
    my ($package) = @_;

    return "package $package;

use strict;
use base 'My::Test::Class';
use Test::Most;

";
}

sub testSub {
    my ($sub) = @_;

    chomp($sub);
    $sub =~ s/\s*sub\s*(.*?) .*/$1/;

    next unless $sub;

    $sub = 'constructor' if $sub eq 'new';

    my $subDef = "sub $sub : Tests() {
    my \$test = shift;

    can_ok(\$test->class(), '$sub');
}

";
    return $subDef;
}

=pod

=head1 NAME

testMaker.pl

=head1 SYNOPSIS

./testMaker.pl

To be run in base dir of project

assuming existing structure:

=over 4

=item lib/My/Class.pm

=item lib/My/Class/Stuff.pm

=back

will create

=over 4

=item t/run.t

Test::Class test runner

=item t/tests/My/Test/Class.pm

Test::Class custom base class

=item t/tests/Test/My/Class.pm

A test class based on lib/My/Class.pm

=item t/tests/Test/My/Class/Stuff.pm

A test class based on lib/My/Class/Stuff.pm

=back

=head1 DESCRIPTION

If you have an existing structure of classes in /lib running this script
will create a skeleton Test::Class test class for each. Within that test
class each of the subs in you lib class will have a test sub created with
a can_ok test.

The test classes will contain test subs, with a use_ok for that sub. The
custom base test class, as parent, will make sure each child test class
on startup performs a use_ok for the lib class you are testing

The script expects to be run in the base of your project (eg the dir
which contains lib/ t/ and so on).

Should a Test::Class test runner not exist one will be created.

A custom base class will also be created if it does not exist, which all
test classes created by this script inherit from.

Once the script has been run, all being well, you will be able to run your
test suite from the base dir of your project using

C<prove -lv --merge>

This script performs a check for L<Test::Class>, L<Test::More> and
L<Class::Data::Inheritable> and will complain at you if they are not
installed. To go ahead and run the script anyway use the -f flag:

C<testMaker.pl -f>

=head1 CAVEATS

It currently does not play nicely with subs which have prototypes, L<which
you probably shouldn't be using anyway!|http://www.perlmonks.org/?node_id=861966>

It currently does not filter out any sub definitions which are contained
within pod sections of your class, and probably never will.

=head1 SEE ALSO

L<Test::Class>

L<Class::Data::Inheritable>

L<Test::More>

L<Test::Most>

=head1 VERSION

version 1.1

=head1 ACKNOWLEDGEMENTS

Based on the excellent tutorial on Test::Class by Curtis "Ovid" Poe,
<ovid at cpan.org> L<found here|http://www.modernperlbooks.com/mt/2009/03/organizing-test-suites-with-testclass.html>

=head1 AUTHOR

Tom Beresford <me at tomberesford dot co dot uk>

=head1 COPYRIGHT AND LICENSE

Copyright 2013 Tom Beresford, All Rights Reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

__DATA__
package My::Test::Class;

use Test::More;
use base qw<Test::Class Class::Data::Inheritable>;

BEGIN {
    __PACKAGE__->mk_classdata('class');
}

sub startup : Tests( startup => 1 ) {
    my $test = shift;

    (my $class = ref $test) =~ s/^Test:://;
    return ok 1, "$class loaded" if $class eq __PACKAGE__;
    use_ok $class or die;
    $test->class($class);
}

1;
